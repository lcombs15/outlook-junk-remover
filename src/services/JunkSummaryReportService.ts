import {JunkEvaluation} from "./junk/JunkService";
import {JunkReport, JunkReportSection} from "../entity/junk-report";
import Email from "../entity/email";

export class JunkSummaryReportService {

    public getReport(evaluations: Array<{ evaluation: JunkEvaluation, email: Email }>): JunkReport {
        const report: JunkReport = {
            junkReport: {},
            notJunkReport: {}
        }

        evaluations.forEach(({evaluation, email}) => {
            const reportEntry = evaluation.isJunk ? report.junkReport : report.notJunkReport;

            reportEntry[evaluation.reason] = reportEntry[evaluation.reason] || [];
            reportEntry[evaluation.reason].push(email)
        })

        return report;
    }

    public printReport(report: JunkReport): void {
        this.printSection('JUNK', report.junkReport);
        this.printSection('NOT JUNK', report.notJunkReport);
    }

    public formatEmailOneLine(email: Email): string {
        return `${email.from.emailAddress.name.trim()} <${email.from.emailAddress.address}> : ${email.subject}`
    }

    private printSection(title: string, section: JunkReportSection): void {
        console.log(`${title} (${this.getSectionSize(section)}):`);
        Object.entries(section).forEach(([reason, emails]) => {
            console.log(`\t${reason} (${emails.length})`);
            emails.forEach(email => {
                console.log(`\t\t+ ${this.formatEmailOneLine(email)}`
                    .replace(/[\r\n]+| {2,}/g, ''))
            })
        })
        console.log(new Array(80).fill('=').join(''))
    }

    private getSectionSize(section: JunkReportSection): number {
        const counts = Object.values(section).map(section => section.length)
        return counts.reduce((sum, count) => sum + count, 0);
    }
}
