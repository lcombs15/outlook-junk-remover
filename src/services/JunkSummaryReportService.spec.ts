import {JunkSummaryReportService} from './JunkSummaryReportService';
import Email from '../entity/email';

describe('JunkSummaryReportService', () => {
    const service = new JunkSummaryReportService();

    describe('formatEmailOneLine', () => {
        it('should format email information into a single line', () => {
            const mockEmail: Email = {
                id: '123',
                receivedDateTime: '2025-08-01T14:25:00Z',
                subject: 'Test Subject',
                from: {
                    emailAddress: {
                        name: 'John Doe',
                        address: 'john.doe@example.com'
                    }
                },
                sender: {
                    emailAddress: {
                        name: 'John Doe',
                        address: 'john.doe@example.com'
                    }
                },
                toRecipients: [],
                ccRecipients: [],
                bccRecipients: [],
                body: {
                    contentType: 'text',
                    content: 'Test content'
                }
            };

            const result = service.formatEmailOneLine(mockEmail);

            expect(result).toMatchSnapshot();
        });

        it('should trim whitespace from sender name', () => {
            const mockEmail: Email = {
                id: '456',
                receivedDateTime: '2025-08-01T14:25:00Z',
                subject: 'Another Test Subject',
                from: {
                    emailAddress: {
                        name: '  Jane Smith  ',
                        address: 'jane.smith@example.com'
                    }
                },
                sender: {
                    emailAddress: {
                        name: 'Jane Smith',
                        address: 'jane.smith@example.com'
                    }
                },
                toRecipients: [],
                ccRecipients: [],
                bccRecipients: [],
                body: {
                    contentType: 'text',
                    content: 'Another test content'
                }
            };

            const result = service.formatEmailOneLine(mockEmail);

            expect(result).toMatchSnapshot();
        });
    });
});
