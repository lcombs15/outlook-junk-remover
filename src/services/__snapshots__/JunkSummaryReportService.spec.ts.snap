// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`JunkSummaryReportService formatEmailOneLine should format email information into a single line 1`] = `"John Doe [john.doe@example.com]: Test Subject"`;

exports[`JunkSummaryReportService formatEmailOneLine should trim whitespace from sender name 1`] = `"Jane Smith [jane.smith@example.com]: Another Test Subject"`;
